### DISCUSSION TITLE
Additional data for IssueList and SeriesList

### DISCUSSION THREAD

**Discussion by mizaki** [bpepple/metron#164] [API Suggestions]
Hi,

I’ve been working on allowing the use of other comic DB information sources for [Comic Tagger](https://github.com/comictagger/comictagger). I have a working version of one for Metron but I have a slight issue with the lack of data for certain API calls. This makes for a rather bare GUI. 

I was wondering if there is any chance of returning some additional fields for some of the API calls? The two options I have as of now are to leave a lot of information blank or, make multiple calls. Obviously, the multiples call within the 30 a minute limit would mean multiple minutes delays for a large result. Not a great user experience.

Here are the calls and the extra information I would like for the GUI:

##
Series search:
https://metron.cloud/api/series/?name={title} (I presume this is the correct way to search for a series name?)

### Current fields for SeriesList:

id
series
modified

### Additional fields requested:

**year_began
publisher
desc
issue_count
image***

*There is no image for series. I notice the website seems to use the image of the first issue. It would be nice to include the first issue image url with the series search.

##
Issue list for series:

https://metron.cloud/api/issue/?series_id={id}

### Current fields for IssueList:

id
issue
cover_date
modified

### Additional fields requested:

**desc
image
number [added in an edit because I missed it]
name***

*issue returns the series name and number which is usable but having the “name” field from the issue would be nicer.

### How it looks currently:
![Metron_search_164855](https://github.com/bpepple/metron/assets/1141189/2edf5e8b-efcc-41e4-936f-a26db470cff9)
![Metron_issues_164855](https://github.com/bpepple/metron/assets/1141189/57eb9cca-52ea-4e31-b8c8-f8f53be0be71)

Thanks.

**Comment by bpepple** [comment]
Hi,

I'm not opposed to adding more information, but I do have a couple of concerns:

- **Potential API breaking changes**: The API has been stable the last 4 years, and I'm hesitant to make any breaking changes since there a lot of tools and users that use the current API, and would need plenty of time to migrate/inform existing users.
- **Implementation**: If there are breaking API changes, it would make sense to add versioning to the API for future changes, which would require a decent amount of developer time.
- **Community**: I'd want to get some input from current users/developers, before making any decisions.
- **Lack of time**: Currently, the amount of free time I've got for a new project is pretty limited, so someone from the community would probably need to step up and implement this.
- **ComicTagger Question**: Does ComicTagger still do image hashing for issue matching? If so, I'd have some questions about the average data consumption per user. Depending of what that is, that could impact our server costs since our CDN's charge for usage over 1 TiB a month.

Regarding your field suggestions, most of them make sense, but I don't think adding the `desc` field is a good idea due to it's potential size (it's a `Django TextField` without a max length).

Anyway, I'll point this discussion to some of the folks in our Matrix rooms later today, and get their thoughts..

**Reply by mizaki** [reply]
Hey, thanks for the reply.

- **Potential API breaking changes:** Obviously it's impossible to know how other people have programmed. I would hope that by only adding additional fields it would require no changes but that may not be the case. 
- **Implementation:** My first thought would be to add a new parameter (?addinfo etc.) that would cause a return of the additional fields. That way everything would remain the same unless requested otherwise.
- **Lack of time:** A forever problem I think :) I would prefer someone who knows what they're doing do it but if not, I can give it a try.
- **ComicTagger Question:** Yes, for automatic matching image hashes are used. All images are cached however like the data itself. All requests from comictagger use the `user-agent: comictagger/+version`. You should see my requests in the web log with such. My hope would be to drive more users to contribute and new users would have to create an account in the first place. If there is a sizeable dump in traffic and more funds were needed, I would hope a few of the new users could offer additional financial support if required.

Re: <desc> field. I would say its use for the series it would be a useful "nice to have", issue it wouldn't be missed.

Thanks again, if you have any additional questions, happy to answer.

**Reply by beville** [reply]
@mizaki  Cool stuff!

@bpepple Related to your valid concern about the bandwidth usage of fetching cover images:

* If I recall correctly, the image hashing works with very small versions of the image, so if there is a way to request a low-res version of the image that might be a nice way to allay concerns and allow ComicTagger to not be a bandwidth hog.
* This is more pie-in-sky idea, but I'd always thought it would be cool if the hashes for the covers could be held in some central DB. Then the images wouldn't need be requested at all (for the matching), just the hashes.  I realize this probably wouldn't a minor thing to not only add the information to the Metron schemas (one hash per cover variant stored), but then also do the work to calculate the hashes for all existing images in the DB, and then also for all new covers, as well as adding a new fields for retrieving said hashes via the API.  But thought I'd throw that idea out there if someone (not me :smile: ) wants to do it.

**Reply by bpepple** [reply]
@beville Regarding your suggestions:

- Unfortunately, we save only the original images and use a Redis key-based cache to serve webp version for the website.
- Now your pie-in-sky idea makes **much** more sense to me as an efficient solution. Having a hash created when an image is uploaded shouldn't be terribly difficult to implement (and hashing existing one would be fairly trivial). A couple of things would need to be decided before implementing, tho:

    1. **Hashing method**: Not sure what ComicTagger uses, but Python's built-in hashing functions (md5, sha256, etc) most likely aren't a great option, but something like [ImageHash](https://pypi.org/project/ImageHash/) provides some decent options. Regardless tho, we would need to coordinate the method between the projects to make this data useful for matching purposes.
    2. **Images to hash**: Should it just be the Primary Cover? Or should it also include Variants? If it's just the Primary Cover adding a ```coverhash``` field to the ```Issue``` model in Django would be needed and updating to the save() and delete() methods, otherwise we would probably need to create another DB table to hold this data. My initial thought is just the Primary Cover, since it's easier to implement and I wonder just how many digital comics have a Variant as it's first image.

Anyway, I think this is a pretty good idea and it's probably worth opening a **Feature Request Bug** on it.

**Reply by mizaki** [reply]
I will tag [@lordwelch](https://github.com/lordwelch) because I've had nothing to do with the hashing.

Some quick notes:

ImageHash is used for the hashing https://github.com/comictagger/comictagger/blob/develop/comictaggerlib/issueidentifier.py.

CT does check variant covers if they are available and the primary doesn't match.

If bandwidth does turns out to be a concern, it should be possible to not use the covers for the auto-tagging process. Of course, this removes the benefit of being more confident in the tagging (and with less data). 

The GUI only loads covers when the series or issue is selected (and then caches it) so I don't believe that would be a major load there for bandwidth.

**Reply by bpepple** [reply]
> ImageHash is used for the hashing https://github.com/comictagger/comictagger/blob/develop/comictaggerlib/issueidentifier.py.
 
That appears to be a custom made image hasher the project created from some info on the net (based on a ***really*** quick read of the code). I was thinking a dedicated library (like ImageHash or similar project) might give more choices.

> If bandwidth does turns out to be a concern, it should be possible to not use the covers for the auto-tagging process. Of course, > this removes the benefit of being more confident in the tagging (and with less data).
> 
> The GUI only loads covers when the series or issue is selected (and then caches it) so I don't believe that would be a major load there for bandwidth.

I think in short term bandwidth won't be an issue, but it's worth giving some thought for future growth.

**Reply by lordwelch** [reply]
> That appears to be a custom made image hasher the project created from some info on the net (based on a really quick read of the code). I was thinking a dedicated library (like ImageHash or similar project) might give more choices.

The hash currently used in ComicTagger is equivalent to ahash (average_hash) from [ImageHash]. [ImageHash] was created around the same time ComicTagger was, it just used a different article as it's base. It may not generate the same hashes because of differences in order of operations.

Doing some back of the hand calculations re-tagging my collection through Metron would take between 3-6GB of bandwidth in images alone (assuming all of my issues are in Metron).

Caching works great when you need to fetch the info a second time, but unless you get a duplicate comic ComicTagger will mostly only cache the search for the series (name only) and the actual series info. Every other url is typically only accessed once. Another issue with caching in ComicTagger is that for the auto-tagging (instead of other operations eg auto-identify, search) it does a search using the publish year and the issue number which usually will not benefit from caching unless it is a duplicate. ComicTagger does cache all images by url.

In regards to implementing, I think having this as a separate service that links hashes to different IDs (Metron, CV, Marvel API) may make more sense, not that it would exclude putting cover hashes in Metron. Which hash to use and if to offer multiple hashes would need some more investigation. Hashes can be stored as a uint64 if space is a concern.

[ImageHash]: https://pypi.org/project/ImageHash/

**Reply by beville** [reply]
Some more blue sky thinking:  How cool would it be to actually have the DB service search solely by the image hash?  

https://stackoverflow.com/questions/54497342/how-can-i-calculate-the-difference-between-two-hashes-in-a-mysql-query

I'm guessing this would probably be too computationally expensive (money and time) for a free (or donation-supported) service to offer. And it might not be really viable without a series name/issue number anyway (too many close matches across _all_ covers).  

But maybe something fun to experiment with once there is a dataset to query against.

**Reply by bpepple** [reply]
@lordwelch My guess is linking ***and*** hashing cover images from the various resources would be quite a bit of work, tho if someone wants to do it that'd be great.

**Reply by bpepple** [reply]
Finally got around to opening a feature request for cover hashing. #170

**Reply by bpepple** [reply]
Did some preliminary testing with ImageHash, and using Perceptual hashing seems to be the best choice based on a greater tolerance of minor modifications. What are other people's thoughts on this?

**Reply by lordwelch** [reply]
I've seen tiny color differences change the hash when using both the Difference Hash and the Perceptual hash. The Average Hash works on much bigger differences but will produce the same hash

**Reply by lordwelch** [reply]
In particularly I've been trying to re-create the hashes from the Python ImageHash library in Go which ends up being difficult because webp as an example stores color as YCbCr and PIL converts it to RGB which doesn't use the same calculation as Go which means that some of the colors have an imperceptible shift but throws off the Difference Hash and the Perceptual Hash by a few points

**Reply by bpepple** [reply]
Hmm, I did some testing on around 50 issues (using Metron Images & Comic Archive covers) using Average & Perceptual Hashes, and Perceptual hashing had **much** better matching results.

That being said, it might be worth using a larger dataset (maybe 1,000 or so) to see if that is accurate conclusion. If I can free up some cycles next week, I'll write a better testing suite to check this and publish the results.

**Comment by bpepple** [comment]
> ### Additional fields requested:
> 
> **desc image number [added in an edit because I missed it] name***
> 
> *issue returns the series name and number which is usable but having the “name” field from the issue would be nicer.

@mizaki  What information are you looking for with the `name` field (issue, series, series + vol or year)?

**Reply by mizaki** [reply]
It's the title of the issue. For example the Story Title on the website such as:`"Sisterhood, Part 3"` (https://metron.cloud/issue/the-uncanny-x-men-1981-510/).

This is what a filled in issue would look like ideally:
![image](https://github.com/bpepple/metron/assets/1141189/6e6990c8-0660-463e-8d41-712f278105a7)

**Reply by bpepple** [reply]
Hmm, definitely nor what I thought you meant. I'm inclined to not add that since I'm not quite sure how that would be used to identify an issue, but maybe I'm missing some info.

What is this window being used for? I don't use gui apps for tasks like tagging, so I'm fairly ignorant on what the work-flow is for CT's gui.

**Reply by mizaki** [reply]
It's a "nice to have" and in no way required for the identification list. As it was first designed around CV's data, I imagine it was what was available and to make it look less bare. As you can see, I've used what was available in the `IssueList` (`issue`) to fill it out.

The flow through the GUI is:
- Search for series
- Select series
- List issues for series (image above)
- Select issue (it will auto-select based on issue number)
- Okay to tag (full information for issue is retrieved)

**Reply by bpepple** [reply]
Ok, did some testing on the changes you're asking for. If I add everything you asked for to the IssueList endpoint it see's around a 279% increase in response size. If only series, number, and image fields are added it's see about 131% increase in response size. Right now I think it make sense to not add the Issue desc & name fields since they aren't necessary for the task of identifying an issue.

Since fields are just being added (for now, I' not removing the semi-redundant `issue` field) there shouldn't be any API breakage, so I'll most likely push these changes out later today, and publish a new version of [Mokkari](https://github.com/Metron-Project/mokkari).

**Reply by bpepple** [reply]
I'm also going to add the **year_began** and **issue_count** fields to the SeriesList endpoint. I'm going to hold off on adding the **image** field for now, since I'll need to do some optimizing of the SQL query on that, since we'll be adding a JOIN on the `Issue` model.

**Reply by mizaki** [reply]
That's great, thanks. I have a couple of queries for Mokkari but I'll put them in its repo.

**Reply by mizaki** [reply]
I also realise it seems I didn't want `name` but `title` (`Collection Title`) that should make more sense. Still not a issue if it's missing but thought I'd clear it up.

**Comment by mizaki** [comment]
Sorry to be *that* guy... digging deeper into creating the talker, I didn't notice that Metron has variant cover images (reading is hard sometimes). Having the URL of the variant images in the IssueList would allow checking and displaying of those (obviously). In the GUI you can flip between them. During an auto-tag I can work around it with an additional call for the full issue (which should be quite limited as it's filter by issue number but if it becomes a problem it can always be removed). It's in no way a game stopper of course, another nice to have.

I also noticed that the series in IssueList doesn't have the series id. I don't need it but I thought it was a little odd. Maybe it is odd to need it? :)

**Reply by bpepple** [reply]
> Sorry to be _that_ guy... digging deeper into creating the talker, I didn't notice that Metron has variant cover images (reading is hard sometimes). Having the URL of the variant images in the IssueList would allow checking and displaying of those (obviously). In the GUI you can flip between them. During an auto-tag I can work around it with an additional call for the full issue (which should be quite limited as it's filter by issue number but if it becomes a problem it can always be removed). It's in no way a game stopper of course, another nice to have.

No worries. I don't think it make a lot sense to add it to the IssueList endpoint. I'm think a better workaround would be to expose a GET method for the Variant Covers table (which only exposes the POST/PUT/PATCH methods currently).

> I also noticed that the series in IssueList doesn't have the series id. I don't need it but I thought it was a little odd. Maybe it is odd to need it? :)

Can't remember the exact reason it's not exposed off the top-of-my-head, but I haven't ran across any use-cases that have needed it.

**Reply by mizaki** [reply]
A theoretical on the series id not being in IssueList, I could see this possibility if you did something like: `https://metron.cloud/api/issue/?number=510&series_name=Uncanny%20x-men` It would take a search of series to get the id to then get the full series info. In that instance, it would remove an additional API call.

**Comment by mizaki** [comment]
If anyone wants to give this a try, an alpha version is [here](https://github.com/mizaki/comictagger-metron-talker).

**Reply by bpepple** [reply]
If I can get some free time I'll give it a try.

**Comment by bpepple** [comment]
> In particularly I've been trying to re-create the hashes from the Python ImageHash library in Go which ends up being difficult because webp as an example stores color as YCbCr and PIL converts it to RGB which doesn't use the same calculation as Go which means that some of the colors have an imperceptible shift but throws off the Difference Hash and the Perceptual Hash by a few points

I suspect the cover hash won't be terribly useful for projects that don't use Python, for exactly the reasons you cite above. Luckily, for tasks like tagging comics, Python is a good choice since a low-level language doesn't provide much benefit (if any).

**Comment by mizaki** [comment]
I would like to raise the issue title/name in IssueList again.

As mentioned before the IssueList uses `display_name` and creates that from the `series title (year) #issue number`. When the full issue information is retrieved there is `title/collection_title` and the array of `name`. As I understand it, `title` is solely used for TPB etc.?

From a user standpoint the issue list will contain only the `display_name` generated title which does no reflect what will be populated once the issue is tagged. CT does fetch the full information when an issue is clicked in the issue window which (will, [issue outstanding](https://github.com/comictagger/comictagger/issues/512)) update that title. However, there is also a free text filter for the title which obviously is not much use unless the `title` or `name` is available for all issues. (And we're not going to do a full issue retrieve for each issue in a series.)

I expect your previous answer stands but I thought I'd bring it up again just in case :)

**Reply by bpepple** [reply]
Yeah, for now my opinion hasn't changed in regard to adding the `Collection Title` & `Story Titles` fields to the `issue_list` endpoint, since it doesn't really help much in identifying the series.

**Comment by anonhacker47** [comment]
Implementing something similar to the comictagger and I have gone through the discussions, and couldn't find any point on not adding publisher to the series_list endpoint (sorry, if i missed it).

I wanted to discuss on adding publisher and series_type (now that collected editions are enabled) to the series_list endpoint.
